File: solutions/java/0015.java
/*
 * Leetcode 15: 3Sum. Java. Medium.
 * Tricky Pointer Problem. O(n^2) runtime and O(n) space complexity.
 * The idea is to use three pointers. Sort nums, then use two pointers, one from first+1 and another from the end inwards.
 * Take care of duplicates. If the sum of the three pointers is equal to 0, add to ArrayList.
 * 3/2/2024 Winston Tsui
*/

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> zeroSums = new ArrayList<>();
        Arrays.sort(nums);

        for (int first = 0; first < nums.length - 2; first++) {
            int second = first + 1;
            int third = nums.length - 1;

            // Skip duplicates for the first number.
            if (first > 0 && nums[first] == nums[first - 1])
                continue;

            // Because nums is sorted at this point, just work our way in from both ends of
            // array with second and third pointers.
            while (second < third) {
                int sum = nums[first] + nums[second] + nums[third];

                if (sum < 0)
                    second++;
                else if (sum > 0)
                    third--;
                else {
                    zeroSums.add(Arrays.asList(nums[first], nums[second], nums[third]));

                    // Skips duplicates that may occur for second and third numbers.
                    while (second < third && nums[second] == nums[second + 1])
                        second++;
                    while (second < third && nums[third] == nums[third - 1])
                        third--;

                    second++;
                    third--;
                }

            }
        }

        return zeroSums;
    }
}

File: solutions/python/0015.py
"""
 * Leetcode 15: 3Sum. Python. Medium.
 * O(N^2) runtime and O(1) space complexity (excluding output)
 * Sort array and use three pointers to find triplets.
 * Skip duplicates to avoid duplicate triplets in the result.
 *
 * Input: nums = [-1,0,1,2,-1,-4]
 * Step 1: Sort array: [-4,-1,-1,0,1,2]
 * Step 2: For i=0, find pairs that sum to 4 (using two pointers)
 * Step 3: Skip duplicates, find valid triplets: [[-1,-1,2],[-1,0,1]]
 *
 * 6/19/2025 Winston Tsui
"""

from typing import List

class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        ans = []
        nums.sort()

        for first in range(len(nums)-2):
            # Skip duplicates for first number
            if first > 0 and nums[first] == nums[first-1]:
                continue

            second = first + 1
            third = n - 1

            while second < third:
                current_sum = nums[first] + nums[second] + nums[third]
                
                if current_sum == 0:
                    ans.append([nums[first], nums[second], nums[third]])
                    
                    # Skip duplicates for second and third numbers
                    while second < third and nums[second] == nums[second+1]:
                        second += 1
                    while second < third and nums[third] == nums[third-1]:
                        third -= 1
                        
                    second += 1
                    third -= 1
                    
                # Adjust pointers based on sum comparison
                elif current_sum < 0:
                    second += 1
                else:
                    third -= 1
                    
        return ans
